From ff563563074094786aa79ab0ee7b5d1ded789a3c Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Thu, 21 Aug 2014 13:05:43 -0400
Subject: [ethernet] qlcnic: Fix loopback diagnostic test

Message-id: <1408626422-3941-60-git-send-email-cdupuis@redhat.com>
Patchwork-id: 88643
O-Subject: [PATCH RHEL 7.1 059/138] qlcnic: Fix loopback diagnostic test
Bugzilla: 1089352
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Manish Chopra <manish.chopra@qlogic.com>

Bug 1089352 (https://bugzilla.redhat.com/show_bug.cgi?id=1089352)

net commit id d9c602f033b00ba360a324c0ee5aa59a6838fb40

o Adapter requires that if the port is in loopback mode no traffic should
  be flowing through that port, so on arrival of Link up AEN, do not advertise
  Link up to the stack until port is out of loopback mode

Signed-off-by: Manish Chopra <manish.chopra@qlogic.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic.h b/drivers/net/ethernet/qlogic/qlcnic/qlcnic.h
index d60c28a..35d4876 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic.h
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic.h
@@ -493,6 +493,7 @@ struct qlcnic_hardware_context {
  struct qlcnic_mailbox *mailbox;
  u8 extend_lb_time;
  u8 phys_port_id[ETH_ALEN];
+ u8 lb_mode;
 };
 
 struct qlcnic_adapter_stats {
@@ -817,6 +818,7 @@ struct qlcnic_mac_vlan_list {
 
 #define QLCNIC_ILB_MODE  0x1
 #define QLCNIC_ELB_MODE  0x2
+#define QLCNIC_LB_MODE_MASK 0x3
 
 #define QLCNIC_LINKEVENT 0x1
 #define QLCNIC_LB_RESPONSE 0x2
diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_hw.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_hw.c
index 1d721ef..a4d6442 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_hw.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_hw.c
@@ -1685,12 +1685,6 @@ int qlcnic_83xx_loopback_test(struct net_device *netdev, u8 mode)
   }
  } while ((adapter->ahw->linkup && ahw->has_link_events) != 1);
 
- /* Make sure carrier is off and queue is stopped during loopback */
- if (netif_running(netdev)) {
-  netif_carrier_off(netdev);
-  netif_tx_stop_all_queues(netdev);
- }
-
  ret = qlcnic_do_lb_test(adapter, mode);
 
  qlcnic_83xx_clear_lb_mode(adapter, mode);
@@ -2122,6 +2116,7 @@ static void qlcnic_83xx_handle_link_aen(struct qlcnic_adapter *adapter,
  ahw->link_autoneg = MSB(MSW(data[3]));
  ahw->module_type = MSB(LSW(data[3]));
  ahw->has_link_events = 1;
+ ahw->lb_mode = data[4] & QLCNIC_LB_MODE_MASK;
  qlcnic_advert_link_change(adapter, link_status);
 }
 
diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_io.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_io.c
index 7312ccd..7ced1c5 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_io.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_io.c
@@ -689,6 +689,10 @@ void qlcnic_advert_link_change(struct qlcnic_adapter *adapter, int linkup)
   adapter->ahw->linkup = 0;
   netif_carrier_off(netdev);
  } else if (!adapter->ahw->linkup && linkup) {
+  /* Do not advertise Link up if the port is in loopback mode */
+  if (qlcnic_83xx_check(adapter) && adapter->ahw->lb_mode)
+   return;
+
   netdev_info(netdev, "NIC Link is up\n");
   adapter->ahw->linkup = 1;
   netif_carrier_on(netdev);
-- 
1.7.1